sim_pblog/sim_pblog/sim_pblog.py

```diff
@@ -0,0 +1,355 @@
+#!/usr/bin/python3
+
+# Copyright 2022 Open Source Robotics Foundation,Inc. and Monterey Bay Aquarium Research Institute
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from datetime import datetime, timedelta
+import gzip
+import math
+import os
+
+from buoy_api import Interface
+from buoy_interfaces.msg import BCRecord
+from buoy_interfaces.msg import PCRecord
+from buoy_interfaces.msg import SCRecord
+from buoy_interfaces.msg import TFRecord
+from buoy_interfaces.msg import XBRecord
+import rclpy
+
+from tf_transformations import euler_from_quaternion
+
+# Controller IDs used as the first value in each record
+BatteryConID = 0
+SpringConID = 1
+PowerConID = 2
+TrefoilConID = 4
+CrossbowID = 5
+
+# WECLogger - duplicate legacy pblog behavior substituting ROS2 topics for CANBus
+# Description:
+#     Init phase: (1) Subscribe to all WEC controllers (e.g., power controller,
+#                     spring controller, etc)
+#                 (2) Create and open CSV output log file, with header line, ready
+#                     for writing log records
+#     Run phase : (1) Controller topic callback functions receive topic messages
+#                 (2) Callback functions write new log records with message data
+#                 (3) Fresh log files are created and old files zipped every hour
+#
+#     Notes     : (1) There is always a trailing comma in all header and
+#                     record sections
+
+
+class WECLogger(Interface):
+
+    def __init__(self):
+        super().__init__('sim_pblog')
+        self.loghome = '/tmp'           # TODO: needs to come from cfg or param
+        self.logdir = None
+        self.logfile = None
+        self.logfilename = None
+        self.logfiletime = datetime.fromtimestamp(0)
+        self.pc_header = ''
+        self.bc_header = ''
+        self.xb_header = ''
+        self.sc_header = ''
+        self.tf_header = ''
+        self.start_time = datetime.now()
+        self.logger_time = self.start_time
+        self.logfileinterval = 60 * 60  # in seconds TODO: needs to come from cfg or param
+        self.logfile_setup()
+
+    def __del__(self):
+        self.close_zip_logfile()
```

Comment:
unfortunately, this doesn't finish in time before garbage collection removes the `open` function (race condition)
```
[sim_pblog-1] KeyboardInterrupt
[sim_pblog-1] Exception ignored in: <function WECLogger.__del__ at 0x7f16a2bdaf80>
[sim_pblog-1] Traceback (most recent call last):
[sim_pblog-1]   File "workspace/install/sim_pblog/lib/python3.10/site-packages/sim_pblog/sim_pblog.py", line 80, in __del__
[sim_pblog-1]   File "workspace/install/sim_pblog/lib/python3.10/site-packages/sim_pblog/sim_pblog.py", line 159, in close_zip_logfile
[sim_pblog-1] NameError: name 'open' is not defined
```
which refers to `open` on line 159 before gzipping the current log file

Reply:
> It happens because the name `open` has already been deleted by the garbage collector before the file handler got the time to use it.

https://stackoverflow.com/a/64690062

Reply:
handled it in my PR targeting your branch